### PR TITLE
Make DOM bindings methods not take raw JSContext pointers. Part 2.

### LINKED ISSUE #20377: Make DOM bindings methods not take raw JSContext pointers
There's a lot of methods that accept a `*mut JSContext` argument, and these methods have to be marked unsafe because callers could pass anything to them, not just a valid JSContext pointer. I propose the following:
* create a `struct JSContext(*mut jsapi::JSContext)` type
* implement `Deref` for this new type so it's easy to obtain the inner context pointer
* add an unsafe `from_ptr` static method that returns a new instance of the type
* make the code generator declare trait methods that accept the new type instead of `*mut JSContext` ([here](https://github.com/servo/servo/blob/797fb553bbfdece69538644da89bb0c937ece997/components/script/dom/bindings/codegen/CodegenRust.py#L5546) and [here](https://github.com/servo/servo/blob/797fb553bbfdece69538644da89bb0c937ece997/components/script/dom/bindings/codegen/CodegenRust.py#L6541); both the type )
* adjust generated code to pass `JSContext::from_ptr(cx)` when necessary ([here](https://github.com/servo/servo/blob/797fb553bbfdece69538644da89bb0c937ece997/components/script/dom/bindings/codegen/CodegenRust.py#L3242-L3245))
* update all of the methods defined in `components/script/dom` that need to be changed
* remove any unnecessary uses of `#[allow(unsafe_code)]` on methods that no longer accept raw pointers

No need to run any automated tests. If it builds, it's good enough for a pull request.

### PR DESCRIPTION
**Pull request by marmeladema**
---

- [ ] `./mach build -d` does not report any errors
- [ ] `./mach test-tidy` does not report any errors
- [ ] These changes fix #20377 

- [ ] There are tests for these changes OR
- [ ] These changes do not require tests because ___

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/23872)

### CODE REVIEW

**Comment by marmeladema** [comment]
@jdm i went forward with removing some of the manual code that deal with raw `JSContext`

File: components/script/dom/promise.rs [hunk 1]
```diff
@@ -50,19 +51,21 @@ pub struct Promise {
 /// Private helper to enable adding new methods to Rc<Promise>.
 trait PromiseHelper {
     #[allow(unsafe_code)]
```
Comment (jdm) [inline components/script/dom/promise.rs]: This annotation can be removed now.

File: components/script/dom/testbinding.rs [hunk 2]
```diff
@@ -943,27 +943,20 @@ impl TestBindingMethods for TestBinding {
     #[allow(unrooted_must_root)]
     #[allow(unsafe_code)]
```
Comment (jdm) [inline components/script/dom/testbinding.rs]: This annotation can be removed now.

File: components/script/dom/bindings/interface.rs [hunk 3]
```diff
@@ -128,8 +129,8 @@ impl InterfaceConstructorBehavior {
 pub type TraceHook = unsafe extern "C" fn(trc: *mut JSTracer, obj: *mut JSObject);
 
 /// Create a global object with the given class.
-pub unsafe fn create_global_object(
-    cx: *mut JSContext,
+pub fn create_global_object(
```
Comment (jdm) [inline components/script/dom/bindings/interface.rs]: We still take a `*const libc::c_void` argument, so this method should still be marked unsafe, unfortunately.

**Review by jdm — CHANGES_REQUESTED** [review]
Looks great!

**Comment by jdm** [comment]
I am so pleased by these changes. I love seeing the unsafe blocks scope the actual pieces of unsafety!

**Comment by marmeladema** [comment]
Fixed your comments and rebased over current master :+1:

**Comment by marmeladema** [comment]
Fixed your comments and rebased over current master :+1:

**Pull request merged**